src/server/cluster/cluster_data.cc

```diff
@@ -0,0 +1,65 @@
+extern "C" {
+#include "redis/crc16.h"
+}
+
+#include <string_view>
+
+#include "cluster_data.h"
+
+namespace dfly {
+
+bool cluster_enabled = false;
+
+// If the key contains the {...} pattern, return only the part between { and }
+std::string_view KeyTag(std::string_view key) {
+  size_t s, e; /* start-end indexes of { and } */
+
+  for (s = 0; s < key.length(); s++)
+    if (key[s] == '{')
+      break;
+
+  // No '{'
+  if (s == key.length())
+    return key;
+
+  // '{' found, Check if we have the corresponding '}'.
+  for (e = s + 1; e < key.length(); e++)
+    if (key[e] == '}')
+      break;
+
+  // No '}'
+  if (e == key.length() || e == s + 1)
+    return key;
+
+  // return only the part between { and }
+  return key.substr(s + 1, e - s - 1);
+}
+
+uint16_t ClusterData::keyHashSlot(std::string_view key) {
+  std::string_view tag = KeyTag(key);
+  return crc16(tag.data(), tag.length()) & kMaxSlotNum;
+}
+
+ClusterData::ClusterData() {
+  AddSlots();
+}
+
+bool ClusterData::AddSlots() {
+  // TODO update logic acording to config
+  // currently add all slots to owned slots
+  std::lock_guard lk{slots_mu_};
+  for (uint16_t slot_id = 0; slot_id <= kMaxSlotNum; ++slot_id) {
```

Comment:
It's not ideal to consume so much RAM for so little information. If there are 3 nodes in the cluster, each will have a map with >5000 entries. Even if we moved to `absl::flat_hash_set` the overhead would still be quite big.
However, since we will need to store all slots (to be able to properly redirect), we'll end up storing 16k entries, each holding multiple strings (IDs and roles).
We could use an interval map here (like [Boost's](https://www.boost.org/doc/libs/1_59_0/libs/icl/doc/html/index.html#boost_icl.introduction)), but perhaps it's not that bad to consume <1MB for cluster related metadata.
We could also do a `flat_hash_map<uint16_t, std::shared_ptr<Instance>>` to have a single instance data allocated.
Just an idea, do whatever you want with it.

Reply:
I am a little bit confused. first you say that using absl::flat_hash_set will be overhead , but than you suggest to use flat_hash_map<uint16_t, std::shared_ptr<Instance>>?

Currently I dont have instance data structure. We will need it for the host in the MOVED error, but lets add it once the config is done. Ok?